### PR TITLE
Add language about (im)mutable properties of services and plans

### LINKED ISSUE #189: Add catalog management text
See: https://docs.cloudfoundry.org/services/managing-service-brokers.html#catalog-validation

I think our spec needs some of this text, in particular the stuff from the "Catalog Validation Behaviors" section.

### PR DESCRIPTION
**Pull request by duglin**
And added info on re-querying a broker's catalog.

Closes #189

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -128,16 +128,33 @@ communications could lead to interoperability issues with other platforms.
...
-Warnings for broker authors:
-
-- Be cautious removing services and plans from their catalogs, as platform marketplaces might have provisioned service instances of these plans. Consider your deprecation strategy.
-- Do not change the ids of services and plans. This action is likely to be evaluated by a platform marketplace as a removal of one plan and addition of another. See above warning about removal of plans.
-
-The following sections describe catalog requests and responses in the Service Broker API.
+The first endpoint that a platform will interact with on the broker is the
+service catalog (`/v2/catalog`). This endpoint returns a list of all services
+available from the broker. Platforms query this endpoint from all brokers
+in order to present an aggregated user-facing catalog.
```
Comment (mattmcneeney) [inline spec.md]: `Platforms query this endpoint in order to present an aggregated user-facing catalog.` is better IMO

Comment (mattmcneeney) [inline spec.md]: ~from~ on

File: spec.md [hunk 2]
```diff
@@ -128,16 +128,33 @@ communications could lead to interoperability issues with other platforms.
...
+MUST NOT change the `id` of that resource across queries.
+
+When a platform recieves different `id` values for the same type of resource,
+even if all of the other metadata in those resources are the exact same, it
+MUST treat them as separate instances of that resource.
+
+Broker authors are expected to be cautious when removing services and plans
+from their catalogs, as platforms might have provisioned service instances of
+these plans. Consider your deprecation strategy.
+
```
Comment (mattmcneeney) [inline spec.md]: Maybe add something in here along the lines of:

`Depending on the changes a platform detects in a brokers' catalog, platforms may restrict the actions that users can perform on existing service instances.`

File: spec.md (hunk 1 above)
Comment (duglin) [inline spec.md]: done

File: spec.md (hunk 2 above)
Comment (duglin) [inline spec.md]: What kind of actions do you see being restricted for existing instances?

Comment (mattmcneeney) [inline spec.md]: In CF, if a broker changes an ID, we won't allow new bindings or new instances of the 'old' service. A new service will appear in the marketplace. I wanted to try and cover the fact that changing IDs will have a large impact on end users.

Comment (duglin) [inline spec.md]: I agree that changing IDs will have the same net effect as deleting the old one. I thought you meant that an existing instance would somehow have new restrictions if there was some other change (other than the ID).  I added a variant of your text, see if you're ok with it.

**Comment by duglin** [comment]
reviews needed

**Comment by mattmcneeney** [comment]
LGTM. I think this will be very useful for broker authors to help them understand how the catalog endpoint is used, and details the issue raised on last weeks call regarding ids and names.

We may still want to do something with naming in the future, but this correctly explains the state of the world today.

**Comment by fmui** [comment]
LGTM

**Comment by angarg12** [comment]
LGTM

**Comment by duglin** [comment]
Just one more review needed: @pmorie @avade @shalako @vaikas-google @zrob

File: spec.md [hunk 3]
```diff
@@ -128,16 +128,35 @@ communications could lead to interoperability issues with other platforms.
...
+from previous queries.
+
+When determining what, if anything, has changed on a broker, the platform
+will use the `id` of the resources (services or plans) as the only immutable
+property and MUST use that to locate the same resource as was returned from
+a previous query. Likewise, if a broker wishes for the same resource to be
+treated as the same resource as one returned from a previous query, then it
+MUST NOT change the `id` of that resource across queries.
+
+When a platform recieves different `id` values for the same type of resource,
```
Comment (pmorie) [inline spec.md]: receives

Comment (duglin) [inline spec.md]: typo: recieves

File: spec.md [hunk 4]
```diff
@@ -128,16 +128,35 @@ communications could lead to interoperability issues with other platforms.
...
+
+Periodically, a platform MAY re-query the service catalog endpoint for a
+broker to see if there are any changes to the list of services. Brokers
+MAY add, remove or modify (metadata, plans, etc.) the list of services
+from previous queries.
+
+When determining what, if anything, has changed on a broker, the platform
+will use the `id` of the resources (services or plans) as the only immutable
+property and MUST use that to locate the same resource as was returned from
+a previous query. Likewise, if a broker wishes for the same resource to be
```
Comment (duglin) [inline spec.md]: awkward reword

**Comment by vaikas-google** [comment]
LGTM

**Pull request merged**